### ISSUE TITLE
[Perf][B300] Evaluate an SM103 persistent ATK forward scan

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/flash-linear-attention#1]
## Hypothesis

The shared ATK forward path used by `chunk_precond_gdn` and `chunk_precond_kda` may be launch- and global-memory-traffic-bound on B300 (SM103). The current implementation launches per-chunk summary, sequential chunk scan, and per-token output kernels; materializes fp32 `a`, `sa`, and `ac`; and reloads token inputs in the output stage. Forward is also recomputed before backward.

No performance improvement has been measured yet. This issue tracks a bounded, evidence-first attempt to determine whether an SM103 persistent scan can remove enough launches and intermediate traffic to improve both public operators.

## Baseline

- Commit: `e47d5d20aeb5989b58a3738b872e7c288a9fb75f`
- Main target: `fla/ops/atk/chunk_atk_fwd.py`
- Consumers: `chunk_precond_gdn` and `chunk_precond_kda`

Before changing the kernel, record full green correctness gates and use `benchmarks.ops.verify --profile` plus representative NCU profiles to measure the three ATK stages' share of end-to-end forward and forward-plus-backward latency. If that share cannot plausibly support the full-op acceptance threshold, record a no-go instead of widening the task.

## Bounded scope

If profiling confirms a material ATK bottleneck:

- Add an exactly SM103-guarded Triton path that processes chunks in sequence, retains the ATK carry for a head-dimension tile in fp32, computes the existing per-token ATK states, and writes `k_precond` directly.
- Support a forward-only mode that may omit `a`, `sa`, and `ac`, and a recompute mode that writes the unchanged `a`/`sa`/`ac` contract consumed by `chunk_atk_bwd`.
- Preserve the existing three-stage implementation for all non-SM103 devices and for any SM103 bucket where profiling demonstrates that the fused path is unsuitable. Shape dispatch is permitted only with per-bucket evidence.
- Add a reusable exact `IS_NVIDIA_SM103` device helper only if the existing helpers cannot express the guard safely.

Do not change ATK recurrence or squash semantics, accumulation dtype, public signatures, the backward kernel, operator references, correctness tolerances, or test parametrization. Do not expand into the larger PGDN/PKDA attention kernels.

## Frozen correctness gate

The existing test files and naive/reference implementations remain unmodified. Fast `--gate-k` runs are iteration signals only; promotion requires both full gates without `-k`:

```bash
python -m benchmarks.ops.verify --op chunk_precond_gdn --base e47d5d20aeb5989b58a3738b872e7c288a9fb75f
python -m benchmarks.ops.verify --op chunk_precond_kda --base e47d5d20aeb5989b58a3738b872e7c288a9fb75f
```

The gate must retain forward and backward checks under NaN memory poisoning, dense and variable-length inputs, partial chunks and non-power-of-two dimensions, fp16/bf16/fp32 coverage where present, initial/final ATK state and gradients, learned `log_atk_scale`, KDA gate/recompute modes, and PGDN GVA/state-layout paths.

## B300 performance evidence

- Compare both registered ops in `fwd` and `fwdbwd` modes over the full repository `SHAPE_CONFIGS` matrix against the frozen baseline.
- Include long-context D64/D128 cases and a D256 case through the public op entry points to expose carry and occupancy limits.
- Measure `precond_gated_deltanet` and `precond_kda` training throughput with identical dense and `--varlen` workloads on the same B300 allocation.
- Use the same inputs, environment, warmup, repetitions, and statistical method for baseline and candidate, interleaving them when practical. Report median, mean, standard deviation, min, p10, p90, and equal-weight geomean rather than best runs.
- Collect NCU full and source profiles for representative dense and packed-varlen D64/D128/D256 cases. Explain any win through summed stage duration, launch count, DRAM/L2 traffic, compute SOL, occupancy, registers/spills, eligible warps, and dominant stalls.
- Record the exact commands, baseline and candidate commits, SM capability, GPU/driver/CUDA/PyTorch/Triton versions, clock/idle evidence, Slurm job IDs, and raw-log/report paths. Keep raw profile artifacts out of git.

## Acceptance criteria

- Both full frozen correctness gates pass with no numeric-flag, tolerance, reference, or test changes.
- The ATK forward/recompute subpipeline improves by at least 1.20x median on the intended B300 regimes.
- Each public op improves by at least 1.05x equal-weight geomean over its B300 target matrix, with no target shape regressing by more than 3% outside measured noise.
- At least one dense and one variable-length model workload improves training throughput by at least 3%; report both public model families and any memory trade-off.
- NCU attributes the result to fewer launches and/or lower global traffic without an unexplained spill or occupancy collapse.

If the bottleneck is not material, the frozen oracle fails, gains are isolated below the full-op thresholds, or serialization/occupancy dominates, record the baseline, attempted direction, gate status, measured bound, and a no-go conclusion rather than adding broader fallbacks or changing numerics.

**Comment by heiheiha798** [comment]
## Cycle 1 result: no-go at the frozen baseline gate

The untouched baseline could not satisfy the required strict correctness prerequisite on B300, so I did not edit the kernel or run performance measurements.

Baseline: `e47d5d20aeb5989b58a3738b872e7c288a9fb75f`

Environment: NVIDIA B300 SXM6 AC (SM103), driver 580.126.09, CUDA 13.0, PyTorch 2.13.0+cu130, Triton 3.7.1, NCU 2025.3.1.

The required PGDN command was started unchanged:

```bash
python -m benchmarks.ops.verify --op chunk_precond_gdn --base e47d5d20aeb5989b58a3738b872e7c288a9fb75f
```

The strict gate failed on the untouched baseline at:

```text
test_chunk[B2-T75-H4-D64-scale1-gln0.01-mask_p0-qk_l2normFalse-torch.float16]
dg diff: 0.000001 ratio: 29.268791
```

The frozen threshold for that comparison is `0.02`. Eight preceding fused-recurrent cases were green, and subsequent sampled chunk cases were green, but this deterministic failure is enough to make the required full command non-green. The job was stopped at 44% after the failure was established, before benchmarking.

The KDA strict command was also started on the same stack. All cases reached through 27% were green, including the first chunk forward/backward case, but it was stopped once the shared PGDN prerequisite had already invalidated the cycle.

I also checked the repository-aligned PyTorch 2.9.1+cu130 / Triton 3.5.1 stack. It is not usable for this target: Triton's bundled `ptxas` exits with status 255 when compiling `sm_103a`.

Evidence was retained in the ignored optimization workspace:

- Slurm job 5052: B300 environment probe
- Slurm array job 5053 task 0: SM103-capable strict PGDN gate
- Slurm array job 5059 task 1: partial strict KDA gate
- Slurm array job 5057 task 0: incompatible Triton 3.5 SM103 check

Per the gate-first contract, `verify.py` did not report a benchmark on the red PGDN gate. No source/test/reference file changed, no performance or NCU claim is made, and no draft PR was created. Loosening the oracle or enabling a test-only relaxation would violate this issue's frozen contract, so this cycle is recorded as no-go rather than widening scope.
